File: ext/doc/Dse/Session.php
<?php

/*
  Copyright (c) 2017 DataStax, Inc.
  This software can be used solely with DataStax Enterprise. Please consult the
  license at http://www.datastax.com/terms/datastax-dse-driver-license-terms
*/

namespace Dse;

/**
 * A class that contains and manages connections to DSE. It allows for
 * executing or preparing queries for execution.
 *
 * @see Cluster::connect()
 * @see Cluster::connectAsync()
 */
interface Session {

    /**
     * Execute a query.
     *
     * @param string|Statement $statement string or statement to be executed.
     * @param array|ExecutionOptions|null $options execution options (optional)
     *
     * @throws Exception
     *
     * @return Rows A collection of rows.
     */
    public function execute($statement, $options);

    /**
     * Execute a query asynchronously. This method returns immediately, but
     * the query continues execution in the background.
     *
     * @param string|Statement $statement string or statement to be executed.
     * @param array|ExecutionOptions|null $options execution options (optional)
     *
     * @return FutureRows A future that can be used to retrieve the result.
     */
    public function executeAsync($statement, $options);

    /**
     * Prepare a query for execution.
     *
     * @param string $cql The query to be prepared.
     * @param ExecutionOptions $options Options to control preparing the query.
     *
     * @throws Exception
     *
     * @return PreparedStatement A prepared statement that can be bound with parameters and executed.
     */
    public function prepare($cql, $options);

    /**
     * Asynchronously prepare a query for execution.
     *
     * @param string $cql The query to be prepared.
     * @param ExecutionOptions $options Options to control preparing the query.
     *
     * @return FuturePreparedStatement A future that can be used to retrieve the prepared statement.
     */
    public function prepareAsync($cql, $options);

    /**
     * Close the session and all its connections.
     *
     * @param double $timeout The amount of time in seconds to wait for the session to close.
     *
     * @throws Exception
     *
     * @return null Nothing.
     */
    public function close($timeout);

    /**
     * Asynchronously close the session and all its connections.
     *
     * @return FutureClose A future that can be waited on.
     */
    public function closeAsync();

    /**
     * Get a snapshot of the cluster's current schema.
     *
     * @return Schema A snapshot of the cluster's schema.
     */
    public function schema();

    /**
     * Execute graph queries.
     *
     * @param string|Graph\Statement $statement A graph statement or query string to be executed.
     * @param array $options Options to control execution of the graph query.
     *
     *                       * array["graph_language"]     string       Graph language; default "gremlin-groovy"
     *                       * array["graph_source"]       string       Graph source; default "g". If running
     *                                                                  analytics (OLAP) query then it should
     *                                                                  use "a"
     *                       * array["graph_name"]         string       Graph name
     *                       * array["read_consistency"]   consistency  Read consistency of graph queries;
     *                                                                  default `Dse::CONSISTENCY_ONE`
     *                       * array["write_consistency"]  consistency  Write consistency of graph queries;
     *                                                                  default `Dse::CONSISTENCY_QUORUM`
     *                       * array["request_timeout"]    double|int   Request time of graph queries in
     *                                                                  seconds; default `0` (no timeout)
     *                       * array["timestamp"]          int|string   Either an integer or integer string
     *                                                                  timestamp that represents the number
     *                                                                  of microseconds since the epoch
     *
     * @throws Exception
     *
     * @return Graph\ResultSet A result set.
     */
    public function executeGraph($statement, $options);

    /**
     * Asynchronously execute a graph query.
     *
     * @param string|Graph\Statement $statement A graph statement or query string to be executed.
     * @param array $options Options to control execution of the graph query.
     *                       @see Session::executeGraph()
     *
     * @return Graph\FutureResultSet A future that can be used to retrieve the result set.
     */
    public function executeGraphAsync($statement, $options);

}

File: ext/src/Session.c
/*
  Copyright (c) 2017 DataStax, Inc.

  This software can be used solely with DataStax Enterprise. Please consult the
  license at http://www.datastax.com/terms/datastax-dse-driver-license-terms
*/

#include "php_driver.h"

zend_class_entry *php_driver_session_ce = NULL;

ZEND_BEGIN_ARG_INFO_EX(arginfo_execute, 0, ZEND_RETURN_VALUE, 1)
  ZEND_ARG_INFO(0, statement)
  PHP_DRIVER_NAMESPACE_ZEND_ARG_OBJ_INFO(0, options, ExecutionOptions, 0)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_prepare, 0, ZEND_RETURN_VALUE, 1)
  ZEND_ARG_INFO(0, cql)
  PHP_DRIVER_NAMESPACE_ZEND_ARG_OBJ_INFO(0, options, ExecutionOptions, 0)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_timeout, 0, ZEND_RETURN_VALUE, 0)
  ZEND_ARG_INFO(0, timeout)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_none, 0, ZEND_RETURN_VALUE, 0)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_execute_graph, 0, ZEND_RETURN_VALUE, 1)
  ZEND_ARG_INFO(0, statement)
  ZEND_ARG_ARRAY_INFO(0, options, 1)
ZEND_END_ARG_INFO()

static zend_function_entry php_driver_session_methods[] = {
  PHP_ABSTRACT_ME(Session, execute, arginfo_execute)
  PHP_ABSTRACT_ME(Session, executeAsync, arginfo_execute)
  PHP_ABSTRACT_ME(Session, prepare, arginfo_prepare)
  PHP_ABSTRACT_ME(Session, prepareAsync, arginfo_prepare)
  PHP_ABSTRACT_ME(Session, close, arginfo_timeout)
  PHP_ABSTRACT_ME(Session, closeAsync, arginfo_none)
  PHP_ABSTRACT_ME(Session, schema, arginfo_none)
  PHP_ABSTRACT_ME(Session, executeGraph, arginfo_execute_graph)
  PHP_ABSTRACT_ME(Session, executeGraphAsync, arginfo_execute_graph)
  PHP_FE_END
};

void php_driver_define_Session(TSRMLS_D)
{
  zend_class_entry ce;

  INIT_CLASS_ENTRY(ce, PHP_DRIVER_NAMESPACE "\\Session", php_driver_session_methods);
  php_driver_session_ce = zend_register_internal_class(&ce TSRMLS_CC);
  php_driver_session_ce->ce_flags |= ZEND_ACC_INTERFACE;
}
